Update description on rule based index selection (#6160)
pingcap/docs: choose-index.md

## Index selection rules

TiDB selects indexes based on rules or cost. the based rules include pre rules and Skyline-pruning. When selecting an index, TiDB tries the pre rule first. If the existing index satisfies a pre rule, TiDB will directly select the index. Otherwise, TiDB will use Skyline-Pruning to exclude unqualified indexes, and then based on the cost estimation of each operator for accessing tables, select the index with the lowest cost.

### Selection based on rules

#### Pre rules

TiDB uses the following heuristic pre rules to select index:

> + rule 1: If the existing index satisfies "unique indexes with full match + no need to retrieve rows from a table (which means the plan generated by the index is IndexReader operator)", directly select this index.

Reviewer: Does `if the existing index satisfies` have the same meaning with `If there exists an index satisfying ....`? I'm not sure.

Author: Sorry, I thought "存在索引” is a term. fixed in https://github.com/pingcap/docs/pull/6160/commits/56651765f33edbdd612dc288e574c67cee6db748.

> + rule 2: If the existing index satisfies "unique indexes + need to retrieve rows from a table (which means the plan generated by the index is IndexReader operator)", select the index with the smallest number of retrieve rows from a table as the candidate index.

Reviewer: with the smallest number of rows to retrieve from a table

Reviewer: or to be retrieved from a table?

Author: Thx ! fixed in https://github.com/pingcap/docs/pull/6160/commits/6eb3baaeeb8c300d3ce32217aca393adb361f40c.

+ rule 3: If an index satisfies "common index + no need to retrieve rows from a table + the number of rows to be read is less than the value of a certain threshold", select the index with the smallest number of rows to be read as the candidate index.

> + rule 4: If only one candidate index is selected based on rule 2 and 3, select this index. If two candidate indexes are separately selected based on rule 2 and 3, select the index with the smaller number of rows to be read (the number of rows with index + the number of rows to be retrieved from a table).

Reviewer 2: ```suggestion
+ Rule 4: If only one candidate index is selected based on rule 2 and 3, select this candidate index. If two candidate indexes are respectively selected based on rule 2 and 3, select the index with the smaller number of rows to be read (the number of rows with index + the number of rows to be retrieved from a table).
```

> + rule 4: If only one candidate index is selected based on rule 2 and 3, select this index. If two candidate index is separately selected based on rule 2 and 3, select the index with the smallest number of rows to be read (the number of rows with index + the number of rows to be retrieved from a table).

Reviewer: If two candidate indexes are separately selected based on rule 2 and 3

Reviewer: Since we compare two indexes, is it `the smaller` rather than `the smallest`?

Author: Thx! fixed in https://github.com/pingcap/docs/pull/6160/commits/ff5bc24bae4a5739cd0b3393e1c08f4ce92f65cb.

> The "index with full match" in the above rules means each indexed column has the equal condition. When executing the `EXPLAIN FORMAT = 'verbose' ...` statement, if the pre rules matches an index, TiDB will output a NOTE level warning indicating that the index matches the pre rule.

Reviewer 2: ```suggestion
The "index with full match" in the above rules means each indexed column has the equal condition. When executing the `EXPLAIN FORMAT = 'verbose' ...` statement, if the pre-rules match an index, TiDB outputs a NOTE level warning indicating that the index matches the pre-rule.
```

> The "unique indexes with full match" in rule 1 means each indexed column has the equivalent qualification. When executing the `EXPLAIN FORMAT = 'verbose' ...` statement, if the pre rules matches an index, TiDB will output a NOTE level warning indicating that the index matches the pre rule.

Author: Previous expression about "索引列" is "indexed column", should I change them all to "index column"?

> The "unique indexes with full match" in rule 1 means each indexed column has the equivalent qualification. When executing the `EXPLAIN FORMAT = 'verbose' ...` statement, if the pre rules matches an index, TiDB will output a NOTE level warning indicating that the index matches the pre rule.

Author: The rest fixed in https://github.com/pingcap/docs/pull/6160/commits/6f02adbdb0c0a60591d1c159f850713352710a69. Thx!

> The "unique indexes with full match" in rule 1 means each indexed column has the equivalent qualification. When executing the `EXPLAIN FORMAT = 'verbose' ...` statement, if the pre rules matches an index, TiDB will output a NOTE level warning indicating that the index matches the pre rule.

Reviewer: > Previous expression about "索引列" is "indexed column", should I change them all to "index column"?

I think there is no need to change. Just keep the original `indexed column` as before.

> In the following example, because the index `idx_b` meets the condition "unique index + need to retrieve rows from a table" in rule 2, TiDB selects the index `idx_b` as the access path, and `SHOW WARNING` returns a note indicating that the index  `idx_b` matches the pre rule.

Reviewer 2: ```suggestion
In the following example, because the index `idx_b` meets the condition "unique index + the need to retrieve rows from a table" in rule 2, TiDB selects the index `idx_b` as the access path, and `SHOW WARNING` returns a note indicating that the index  `idx_b` matches the pre-rule.
```

```sql
mysql> CREATE TABLE t(a INT PRIMARY KEY, b INT, c INT, UNIQUE INDEX idx_b(b));
Query OK, 0 rows affected (0.01 sec)

> mysql> EXPLAIN FORMAT = 'verbose' SELECT b, c FROM t WHERE b = 3 OR b = 6;

Reviewer 2: ```suggestion

mysql> EXPLAIN FORMAT = 'verbose' SELECT b, c FROM t WHERE b = 3 OR b = 6;
```

+-------------------+---------+---------+------+-------------------------+------------------------------+
| id                | estRows | estCost | task | access object           | operator info                |
+-------------------+---------+---------+------+-------------------------+------------------------------+
| Batch_Point_Get_5 | 2.00    | 8.80    | root | table:t, index:idx_b(b) | keep order:false, desc:false |
+-------------------+---------+---------+------+-------------------------+------------------------------+
1 row in set, 1 warning (0.00 sec)

> mysql> SHOW WARNINGS;

Reviewer 2: ```suggestion

mysql> SHOW WARNINGS;
```

+-------+------+-------------------------------------------------------------------------------------------+
| Level | Code | Message                                                                                   |
+-------+------+-------------------------------------------------------------------------------------------+
| Note  | 1105 | unique index idx_b of t is selected since the path only has point ranges with double scan |
+-------+------+-------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### Skyline-pruning

Skyline-pruning is a heuristic filtering rule for indexes, which can reduce the probability of wrong index selection caused by wrong estimation. To judge an index, the following three dimensions are needed:

> - Whether it needs to retrieve rows from a table when you select the index to access the table (that is, the plan generated by the index is IndexReader operator or IndexLookupReader operator). Indexes that do not retrieve rows from a table are better on this dimension than indexes that do. If both indexes need to retrieve rows, compare how many filter conditions are covered by the indexed columns. Filter conditions mean the `where` condition that can be judged based on the index. If the column set of an index covers more access conditions, the smaller the number of retrieved rows from a table, and the better the index is in this dimension.

Reviewer 2: ```suggestion
- Whether it needs to retrieve rows from a table when you select the index to access the table (that is, the plan generated by the index is IndexReader operator or IndexLookupReader operator). Indexes that do not retrieve rows from a table are better on this dimension than indexes that do. If both indexes need TiDB to retrieve rows from the table, compare how many filtering conditions are covered by the indexed columns. Filtering conditions mean the `where` condition that can be judged based on the index. If the column set of an index covers more access conditions, the smaller the number of retrieved rows from a table, and the better the index is in this dimension.
```

- Select whether the index satisfies a certain order. Because index reading can guarantee the order of certain column sets, indexes that satisfy the query order are superior to indexes that do not satisfy on this dimension.

> - How many access conditions are covered by the indexed columns. An "access condition" is a where condition that can be converted to a column range. And the more access conditions an indexed column set covers, the better it is in this dimension.
> 

Reviewer 2: ```suggestion
```

> For these three dimensions, if the index `idx_a` is not worse than the index `idx_b` in all three dimensions and one of the dimensions is better than `idx_b`, then `idx_a` is preferred. When executing the `EXPLAIN FORMAT = 'verbose' ...` statement, if skyline-pruning excludes some indexes, TiDB will output a NOTE level warning listing the reserved indexes after skyline-pruning's exclusion.

Reviewer 2: ```suggestion
For these three dimensions above, if the index `idx_a` performs no worse than the index `idx_b` in all three dimensions and performs better than `idx_b` in one dimension, then `idx_a` is preferred. When executing the `EXPLAIN FORMAT = 'verbose' ...` statement, if skyline-pruning excludes some indexes, TiDB outputs a NOTE level warning listing the remaining indexes after the skyline-pruning exclusion.
```

> In the following example, the index `idx_b` and `idx_e` are both inferior to `idx_b_c`, so they are excluded by skyline-pruning. The returned result of `SHOW WARNING` displays the remaining indexes after skyline-pruning.

Reviewer 2: ```suggestion
In the following example, the indexes `idx_b` and `idx_e` are both inferior to `idx_b_c`, so they are excluded by skyline-pruning. The returned result of `SHOW WARNING` displays the remaining indexes after skyline-pruning.
```

```sql
mysql> CREATE TABLE t(a INT PRIMARY KEY, b INT, c INT, d INT, e INT, INDEX idx_b(b), INDEX idx_b_c(b, c), INDEX idx_e(e));
Query OK, 0 rows affected (0.01 sec)

> mysql> EXPLAIN FORMAT = 'verbose' SELECT * FROM t WHERE b = 2 AND c > 4;

Reviewer 2: ```suggestion

mysql> EXPLAIN FORMAT = 'verbose' SELECT * FROM t WHERE b = 2 AND c > 4;
```

+-------------------------------+---------+---------+-----------+------------------------------+----------------------------------------------------+
| id                            | estRows | estCost | task      | access object                | operator info                                      |
+-------------------------------+---------+---------+-----------+------------------------------+----------------------------------------------------+
| IndexLookUp_10                | 33.33   | 738.29  | root      |                              |                                                    |
| ├─IndexRangeScan_8(Build)     | 33.33   | 2370.00 | cop[tikv] | table:t, index:idx_b_c(b, c) | range:(2 4,2 +inf], keep order:false, stats:pseudo |
| └─TableRowIDScan_9(Probe)     | 33.33   | 2370.00 | cop[tikv] | table:t                      | keep order:false, stats:pseudo                     |
+-------------------------------+---------+---------+-----------+------------------------------+----------------------------------------------------+
3 rows in set, 1 warning (0.00 sec)

> mysql> SHOW WARNINGS;

Reviewer 2: ```suggestion

mysql> SHOW WARNINGS;
```

## Index selection rules

TiDB selects indexes based on rules or cost. The based rules include pre-rules and skyline-pruning. When selecting an index, TiDB tries the pre-rule first. If an index satisfies a pre-rule, TiDB directly selects this index. Otherwise, TiDB uses skyline-pruning to exclude unsuitable indexes, and then selects the index with the lowest cost based on the cost estimation of each operator that accesses tables.

### Rule-based selection

#### Pre-rules

> TiDB uses the following heuristic pre rules to select indexes:

Reviewer 2: ```suggestion
TiDB uses the following heuristic pre-rules to select indexes:
```

+-------------------+---------+---------+------+-------------------------+------------------------------+
| Batch_Point_Get_5 | 2.00    | 8.80    | root | table:t, index:idx_b(b) | keep order:false, desc:false |
+-------------------+---------+---------+------+-------------------------+------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> SHOW WARNINGS;
+-------+------+-------------------------------------------------------------------------------------------+
| Level | Code | Message                                                                                   |
+-------+------+-------------------------------------------------------------------------------------------+
| Note  | 1105 | unique index idx_b of t is selected since the path only has point ranges with double scan |
+-------+------+-------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### Skyline-pruning

Skyline-pruning is a heuristic filtering rule for indexes, which can reduce the probability of wrong index selection caused by wrong estimation. To judge an index, the following three dimensions are needed:

> 

Reviewer 2: @Liuxiaozhen12 Looks like "- 索引的列涵盖了多少访问条件。“访问条件”指的是可以转化为某列范围的 `where` 条件，如果某个索引的列集合涵盖的访问条件越多，那么它在这个维度上更优。" is missing. Could you help translate the sentences? Thanks.

https://github.com/pingcap/docs-cn/pull/6815/files#diff-6e8b2a27166b0213a4f5283b2f20c90a9163c95ed74cda9943a819b357e3dbfbR77

| ├─IndexRangeScan_8(Build)     | 33.33   | 2370.00 | cop[tikv] | table:t, index:idx_b_c(b, c) | range:(2 4,2 +inf], keep order:false, stats:pseudo |
| └─TableRowIDScan_9(Probe)     | 33.33   | 2370.00 | cop[tikv] | table:t                      | keep order:false, stats:pseudo                     |
+-------------------------------+---------+---------+-----------+------------------------------+----------------------------------------------------+
3 rows in set, 1 warning (0.00 sec)

mysql> SHOW WARNINGS;
+-------+------+------------------------------------------------------------------------------------------+
| Level | Code | Message                                                                                  |
+-------+------+------------------------------------------------------------------------------------------+
| Note  | 1105 | [t,idx_b_c] remain after pruning paths for t given Prop{SortItems: [], TaskTp: rootTask} |
+-------+------+------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

> ### Selection based on cost estimation

Reviewer 2: ```suggestion
### Cost estimation-based selection
```

+-------------------+---------+---------+------+-------------------------+------------------------------+
| Batch_Point_Get_5 | 2.00    | 8.80    | root | table:t, index:idx_b(b) | keep order:false, desc:false |
+-------------------+---------+---------+------+-------------------------+------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> SHOW WARNINGS;
+-------+------+-------------------------------------------------------------------------------------------+
| Level | Code | Message                                                                                   |
+-------+------+-------------------------------------------------------------------------------------------+
| Note  | 1105 | unique index idx_b of t is selected since the path only has point ranges with double scan |
+-------+------+-------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### Skyline-pruning

Skyline-pruning is a heuristic filtering rule for indexes, which can reduce the probability of wrong index selection caused by wrong estimation. To judge an index, the following three dimensions are needed:

> 

Reviewer 2: ```suggestion

- How many access conditions are covered by the indexed columns. An “access condition” is a where condition that can be converted to a column range. And the more access conditions an indexed column set covers, the better it is in this dimension.

```

## Index selection rules

> TiDB selects indexes based on rules or cost. the based rules include pre rules and Skyline-pruning. When selecting an index, TiDB tries the pre rule first. If an index satisfies a pre rule, TiDB directly selects the index. Otherwise, TiDB uses Skyline-pruning to exclude unqualified indexes, and then based on the cost estimation of each operator for accessing tables, selects the index with the lowest cost.

Reviewer: `the based rules` -> `The based rules`
`Skyline-pruning` -> `skyline-pruning` it seems that we don't capitalize the first letter before the PR

Reviewer: I'm not sure whether `unqualified` corresponds to `不合适`. Maybe we can also consider `improper` or `unsuitable`?

Author: `unsuitable` sounds better. fixed in https://github.com/pingcap/docs/pull/6160/commits/40a51d14d5d26fe58f28df41c5497e8902a58888.
